### PR TITLE
Fix undo log

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR enables undo log testing as well as fixing memory handling issues (double freeing attribute sets)

### CODE REVIEW

**Review by AviAvni — APPROVED** [review]

File: src/graph/entities/attribute_set.c [hunk 1]
```diff
@@ -149,6 +143,41 @@ void AttributeSet_Add
...
+// adds an attribute to the set
+void AttributeSet_Add
+(
+	AttributeSet *set,     // set to update
+	Attribute_ID attr_id,  // attribute identifier
+	SIValue value          // attribute value
+) {
+	// validate value type
+	// value must be a valid property type
+	ASSERT(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE);
```
Comment (AviAvni) [inline src/graph/entities/attribute_set.c:179]: this is common to both add function why it is not in the prepare?

File: src/undo_log/undo_log.c [hunk 2]
```diff
@@ -78,12 +99,12 @@ static void _UndoLog_Rollback_Update_Entity
 
 		// update indices
 		if(update_op->entity_type == GETYPE_NODE) {
-			Graph_UpdateEntity((GraphEntity *)&update_op->n, update_op->attr_id,
-				update_op->orig_value, update_op->entity_type);
+			_UndoLog_Restore_Entity_Property((GraphEntity *)&update_op->n, update_op->attr_id,
```
Comment (AviAvni) [inline src/undo_log/undo_log.c:102]: in this case that we added new function in the undo log we can just sent the update_op as parameter no?

**Review by AviAvni — CHANGES_REQUESTED** [review]

File: src/graph/entities/attribute_set.c (hunk 1 above)
Comment (DvirDukhan) [reply src/graph/entities/attribute_set.c:179]: It is just an assertion on the value. The prepare function does some actual logic.

File: src/undo_log/undo_log.c (hunk 2 above)
Comment (DvirDukhan) [reply src/undo_log/undo_log.c:102]: I prefer to leave it this way as we will have to infer again the entity type

**Review by AviAvni — APPROVED** [review]

**Pull request merged**